lunatrace/bsl/backend/src/snapshot/package-tree.ts

```diff
@@ -0,0 +1,165 @@
+      const { id, dependency, parentId } = entry;
+      if (!dependency.version || !dependency.name) {
+        return null;
+      }
+
+      return {
+        /*
+         TODO (cthompson) we are generating the ids for the database here. This avoids a roundtrip from the server
+         so we can have an id on hand for `depended_by_relationship_id` but could result in an insertion error.
+         What is the best way to recover from this? Theoretically, on error, we can
+         regenerate the graphql insertion query which would create new UUIDs which would be incredibly unlikely to have another
+         collision. For now, this is most likely fine and shouldn't result in any issues, the planets need to align for this to
+         be a problem atm.
+         */
+        id,
+        build_id: buildId,
+        depended_by_relationship_id: parentId,
+        release: {
+          data: {
+            version: dependency.version,
+            package: {
+              data: {
+                name: dependency.name,
+                package_manager: 'npm',
+                // TODO (cthompson) theoretically we have the information to fill this in, how do we get this from a PkgTree?
+                custom_registry: '',
+              },
+              on_conflict: {
+                constraint: Package_Constraint.PackagePackageManagerCustomRegistryNameIdx,
+                update_columns: [
+                  Package_Update_Column.Name,
+                  Package_Update_Column.PackageManager,
+                  Package_Update_Column.CustomRegistry,
+                ],
+              },
+            },
+          },
+          on_conflict: {
+            constraint: Package_Release_Constraint.ReleasePackageIdVersionIdx,
+            update_columns: [Package_Release_Update_Column.Version],
+          },
+        },
+        labels: dependency.labels,
+        range: dependency.range,
+      };
+    })
+    .filter(notEmpty);
+}
+
+export async function snapshotPinnedDependencies(buildId: string, repoDir: string) {
+  const pkgTrees = await collectPackageTreesFromDirectory(repoDir);
+
+  const buildDependencyRelationships = pkgTrees
+    .map((pkgTree) => mapPackageTreeToBuildDependencyRelationships(buildId, pkgTree))
+    .flat();
+
+  log.info('generated package tree for pinned dependencies', {
+    count: buildDependencyRelationships.length,
+  });
+
+  const chunkSize = 5000;
+  for (let i = 0; i < buildDependencyRelationships.length; i += 1) {
+    if (i % chunkSize !== 0) {
+      continue;
+    }
+
+    const chunk = buildDependencyRelationships.slice(i, i + chunkSize);
+    log.info('inserting build dependency relationships', {
+      chunkSize: chunk.length,
+    });
+
+    const resp = await hasura.InsertBuildDependencyRelationships({
+      objects: chunk,
+      on_conflict: {
+        constraint: Build_Dependency_Relationship_Constraint.BuildDependencyRelationshipPkey,
```

Comment:
This in particular I would leave out because now if there ever IS a collision, it will just fail silently, change the pkey and not anything else, and maybe even try to break some other foreign keys creating a weird error.  

Do we actually need this for anything? Isn't it better to outright fail?